src/components/ModelingSidebar/ModelingPanes/CodeEditor.tsx

```diff
@@ -0,0 +1,140 @@
+import React, { useEffect, useMemo, useRef, useState } from 'react'
+import {
+  EditorState,
+  EditorStateConfig,
+  Extension,
+  StateEffect,
+} from '@codemirror/state'
+import { EditorView } from '@codemirror/view'
+import { oneDark } from '@codemirror/theme-one-dark'
+
+//reference: https://github.com/sachinraja/rodemirror/blob/main/src/use-first-render.ts
+const useFirstRender = () => {
+  const firstRender = useRef(true)
+
+  useEffect(() => {
+    firstRender.current = false
+  }, [])
+
+  return firstRender.current
+}
+
+const defaultLightThemeOption = EditorView.theme(
+  {
+    '&': {
+      backgroundColor: '#fff',
+    },
+  },
+  {
+    dark: false,
+  }
+)
+
+interface ICodeEditor {
+  onView?: (view: EditorView | null) => void
+  initialDocValue?: EditorStateConfig['doc']
+  extensions?: Extension
+  theme: 'light' | 'dark'
+  autoFocus?: boolean
+  selection?: EditorStateConfig['selection']
+}
+
+interface UseCodeMirror extends ICodeEditor {
+  container?: HTMLDivElement | null
+}
+
+const CodeEditor: React.FC<ICodeEditor> = ({
+  onView,
+  extensions = [],
+  initialDocValue,
+  theme,
+  autoFocus = false,
+  selection,
+}) => {
+  const editorRef = useRef<HTMLDivElement>(null)
+
+  useCodeMirror({
+    container: editorRef.current,
+    onView,
+    extensions,
+    initialDocValue,
+    theme,
+    autoFocus,
+    selection,
+  })
+
+  return <section ref={editorRef}></section>
+}
+
+export function useCodeMirror(props: UseCodeMirror) {
+  const {
+    onView,
+    extensions = [],
+    initialDocValue,
+    theme,
+    autoFocus = false,
+    selection,
+  } = props
+
+  const [container, setContainer] = useState<HTMLDivElement | null>()
+  const [editorView, setEditorView] = useState<EditorView | null>(null)
+
+  const isFirstRender = useFirstRender()
+
+  const targetExtensions = useMemo(() => {
+    let exts = Array.isArray(extensions) ? extensions : []
+    if (theme === 'dark') {
+      exts = [...exts, oneDark]
+    } else if (theme === 'light') {
+      exts = [...exts, defaultLightThemeOption]
+    }
+
+    return exts
+  }, [extensions, theme])
+
+  useEffect(() => {
+    if (isFirstRender || !editorView) return
+    console.log('reconfigure', targetExtensions)
+
+    editorView.dispatch({
+      effects: StateEffect.reconfigure.of(targetExtensions),
+    })
+  }, [targetExtensions])
+
+  useEffect(() => setContainer(props.container), [props.container])
+
+  useEffect(() => {
+    if (container === null) return
+
+    const view = new EditorView({
+      state: EditorState.create({
+        doc: initialDocValue,
+        extensions: [...Array.of(extensions)],
+        selection,
+      }),
+      parent: container,
+    })
+
+    setEditorView(view)
```

Comment:
I can forsee us wanting to access this view outside of the state local to this component, so at some point we might need to add Xstate context or similar, but if this is working fine now, no need to worry about it, I guess we didn't need to pass around the view with the old code-mirror.

Reply:
well we already do this today :) its in the editor manager